### PR TITLE
[New Task] Web App for Containers

### PR DESCRIPTION
**Pull request by SumiranAgg**

### CODE REVIEW

File: Tasks/AzureWebAppOnContainerDeploymentV1/Tests/AzureRmWebAppDeploymentProviderL0Tests.ts [hunk 1]
```diff
@@ -0,0 +1,68 @@
+import tl = require('vsts-task-lib');
+import { AzureRmWebAppDeploymentProvider } from '../azurermwebappdeploymentprovider'
```
Comment (RoshanKumarMicrosoft) [inline Tasks/AzureWebAppOnContainerDeploymentV1/Tests/AzureRmWebAppDeploymentProviderL0Tests.ts]: Missing semicolon

File: Tasks/AzureWebAppOnContainerDeploymentV1/Tests/package-lock.json [hunk 2]
```diff
@@ -0,0 +1,124 @@
+{
```
Comment (RoshanKumarMicrosoft) [inline Tasks/AzureWebAppOnContainerDeploymentV1/Tests/package-lock.json]: Can you please make sure each specific version of module must be in thirdpartyNotice .

File: Tasks/AzureWebAppOnContainerDeploymentV1/azurermwebappdeploymentprovider.ts [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+import { ContainerBasedDeploymentUtility } from 'azurermdeploycommon/operations/ContainerBasedDeploymentUtility';
+
+export class AzureRmWebAppDeploymentProvider{
+    protected taskParams:TaskParameters;
+    protected appService: AzureAppService;
+    protected kuduService: Kudu;
+    protected appServiceUtility: AzureAppServiceUtility;
+    protected kuduServiceUtility: KuduServiceUtility;
+    protected azureEndpoint: AzureEndpoint;
+    protected activeDeploymentID;
```
Comment (RoshanKumarMicrosoft) [inline Tasks/AzureWebAppOnContainerDeploymentV1/azurermwebappdeploymentprovider.ts]: missing type

File: Tasks/AzureWebAppOnContainerDeploymentV1/azurermwebappdeploymentprovider.ts [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+        this.kuduService = await this.appServiceUtility.getKuduService();
+        this.kuduServiceUtility = new KuduServiceUtility(this.kuduService);
+        tl.debug(`Resource Group: ${this.taskParams.ResourceGroupName}`);
+    }
+
+    public async DeployWebAppStep() {
+        tl.debug("Performing container based deployment.");
+
+        let containerDeploymentUtility: ContainerBasedDeploymentUtility = new ContainerBasedDeploymentUtility(this.appService);
+        await containerDeploymentUtility.deployWebAppImage(this.taskParams);
```
Comment (RoshanKumarMicrosoft) [inline Tasks/AzureWebAppOnContainerDeploymentV1/azurermwebappdeploymentprovider.ts]: No much sure about the logic behind "We are waiting for each method to be executed before moving to next statement then what is logic of async method".  Why not sync ? We generally use Async method when we really don't main thread to block on it or context switch between thread to take care of it.

File: Tasks/AzureWebAppOnContainerDeploymentV1/Tests/AzureRmWebAppDeploymentProviderL0Tests.ts (hunk 1 above)
Comment (SumiranAgg) [reply Tasks/AzureWebAppOnContainerDeploymentV1/Tests/AzureRmWebAppDeploymentProviderL0Tests.ts]: Added

File: Tasks/AzureWebAppOnContainerDeploymentV1/Tests/package-lock.json (hunk 2 above)
Comment (SumiranAgg) [reply Tasks/AzureWebAppOnContainerDeploymentV1/Tests/package-lock.json]: Sure

File: Tasks/AzureWebAppOnContainerDeploymentV1/azurermwebappdeploymentprovider.ts (hunk 3 above)
Comment (SumiranAgg) [reply Tasks/AzureWebAppOnContainerDeploymentV1/azurermwebappdeploymentprovider.ts]: Added

File: Tasks/AzureWebAppOnContainerDeploymentV1/taskparameters.ts [hunk 5]
```diff
@@ -0,0 +1,35 @@
...
+    }
+}
+
+export interface TaskParameters {
+    connectedServiceName: string;
+    WebAppName: string;
+    AppSettings?: string;
+    StartupCommand?: string;
+    ConfigurationSettings?: string;
+    ImageName?: string;
```
[the remark this answers is not in the document]
Comment (SumiranAgg) [reply Tasks/AzureWebAppOnContainerDeploymentV1/taskparameters.ts]: Fixed

File: Tasks/AzureWebAppOnContainerDeploymentV1/make.json [hunk 6]
```diff
@@ -0,0 +1,27 @@
+{
+    "common": [
+        {
```
[the remark this answers is not in the document]
Comment (SumiranAgg) [reply Tasks/AzureWebAppOnContainerDeploymentV1/make.json]: Will add this when I clean resources from task.json

File: Tasks/AzureWebAppOnContainerDeploymentV1/azurermwebappdeploymentprovider.ts (hunk 4 above)
Comment (SumiranAgg) [reply Tasks/AzureWebAppOnContainerDeploymentV1/azurermwebappdeploymentprovider.ts]: You are correct we are doing asynchronous calls sequentially. But there is a difference with synchronous calls. With sync we block the thread until response is returned. With aync/await construct, the call returns a promise instantaneously but waits for it to settle down. Thus the thread is not blocked, and we still have a responsive UI. await ensures that the rest of the async function is executed once awaitable is completed.

Comment (RoshanKumarMicrosoft) [reply Tasks/AzureWebAppOnContainerDeploymentV1/azurermwebappdeploymentprovider.ts]: Here we are not using any UI and it is background task.

**Pull request merged**